charts/scalardl/templates/ledger/deployment.yaml

```diff
@@ -29,34 +31,40 @@ spec:
       securityContext:
         {{- toYaml .Values.ledger.podSecurityContext | nindent 8 }}
       volumes:
-      {{- if .Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled }}
+      {{- if or (.Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled) (lookup "v1" "Secret" .Release.Namespace .Values.ledger.scalarLedgerConfiguration.secretName) }}
```

Comment:
In the Ledger charts, we need to add new condition (if statement) to mount the key file that will be set to `scalar.dl.ledger.proof.private_key_path`.

If we use customized `ledger.properties`, we don't set `ledger.scalarLedgerConfiguration.ledgerProofEnabled` in the custom values file. We set `scalar.dl.ledger.proof.enabled=true` in the `ledger.ledgerProperties` directly instead of `ledger.scalarLedgerConfiguration.ledgerProofEnabled`.

However, in the above case, the following condition returns `false` and the key file is not mounted to the Pod even if we enable ledger proof.

```
{{- if .Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled }}
```

We can avoid this issue by using the following custom values file.

```yaml
ledger:

  scalarLedgerConfiguration:
    ledgerProofEnabled: true

  ledgerProperties: |
    ...
    scalar.dl.ledger.proof.enabled=true
    ...
```

However, I think this way is not good since this custom values file has redundancy settings. It set `scalar.dl.ledger.proof.enabled=true` twice.
(Also, we will face the same issue when we will deprecate the `ledger.scalarLedgerConfiguration` in the future.)

So, I think it is better to use another way to avoid this issue.
And, to resolve this issue, I chose [lookup](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function) function of Helm.

The `lookup` function can be used to look up resources in a running Kubernetes cluster.
The following `if` statement checks the specific named `Secret` resource exists or not.
(`Secret` name is set in the custom values file or values.yaml file.)

```
{{- if (lookup "v1" "Secret" .Release.Namespace .Values.ledger.scalarLedgerConfiguration.secretName) }}
```

If the specified named `Secret` (includes key file for Ledger) exists, this chart mounts it to the Pod.
If the specified named `Secret` does NOT exist, this chart doesn't mount anything. (If we set `scalar.dl.ledger.proof.enabled=true` and there is no `Secret`, the Ledger pods will be failed.)

Also, to keep backward-compatibility, I leave the old condition and combine these two conditions using `or` operator.

Reply:
If customized ledger.properties is used, a modified key path may be specified as follows:

```
scalar.dl.ledger.proof.enabled=true
scalar.dl.ledger.proof.private_key_path=/tmp/private.key
```

It seems simpler to have `.Values.ledger.scalarLedgerConfiguration` setting only work with the default `ledger.properties`. In your suggestion above, although you are using customized ledger.properties, the Secret object name needs to match `.Values.ledger.scalarLedgerConfiguration.secretName` and the key name must be `private-key`.

For example, how about allowing volumes and volumeMounts to be customizable and have them used as shown below? This way, if in the future you decide to use credentials other than keys as files in customized ledger.properties, you can do so without changing the chart.

```diff
diff --git a/charts/scalardl/templates/ledger/deployment.yaml b/charts/scalardl/templates/ledger/deployment.yaml
index 88b28d7..6c031af 100644
--- a/charts/scalardl/templates/ledger/deployment.yaml
+++ b/charts/scalardl/templates/ledger/deployment.yaml
@@ -31,7 +31,7 @@ spec:
       securityContext:
         {{- toYaml .Values.ledger.podSecurityContext | nindent 8 }}
       volumes:
-      {{- if or (.Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled) (lookup "v1" "Secret" .Release.Namespace .Values.ledger.scalarLedgerConfiguration.secretName) }}
+      {{- if .Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled }}
         - name: "{{ .Values.ledger.scalarLedgerConfiguration.secretName }}"
           secret:
             secretName: "{{ .Values.ledger.scalarLedgerConfiguration.secretName }}"
@@ -39,6 +39,9 @@ spec:
         - name: scalardl-ledger-properties-volume
           configMap:
             name: {{ include "scalardl.fullname" . }}-ledger-properties
+        {{- with .Values.ledger.volumes }}
+          {{- toYaml . nindent 8 }}
+        {{- end }}
       containers:
         - name: {{ .Chart.Name }}-ledger
           securityContext:
@@ -46,7 +49,7 @@ spec:
           image: "{{ .Values.ledger.image.repository }}:{{ .Values.ledger.image.version }}"
           imagePullPolicy: {{ .Values.ledger.image.pullPolicy }}
           volumeMounts:
-          {{- if or (.Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled) (lookup "v1" "Secret" .Release.Namespace .Values.ledger.scalarLedgerConfiguration.secretName) }}
+          {{- if .Values.ledger.scalarLedgerConfiguration.ledgerProofEnabled }}
             - name: "{{ .Values.ledger.scalarLedgerConfiguration.secretName }}"
               mountPath: "/keys"
               readOnly: true
@@ -54,6 +57,9 @@ spec:
             - name: scalardl-ledger-properties-volume
               mountPath: /scalar/ledger/ledger.properties.tmpl
               subPath: ledger.properties.tmpl
+            {{- with .Values.ledger.volumeMounts }}
+              {{- toYaml . | nindent 12 }}
+            {{- end }}
           ports:
           - containerPort: 50051
           - containerPort: 50052
diff --git a/charts/scalardl/values.yaml b/charts/scalardl/values.yaml
index 3dcf184..8096758 100644
--- a/charts/scalardl/values.yaml
+++ b/charts/scalardl/values.yaml
@@ -210,6 +210,9 @@ ledger:
     # runAsNonRoot: true
     # runAsUser: 1000

+  volumeMounts: []
+  volumes: []
+
   service:
     # -- service types in kubernetes
     type: ClusterIP
```